source/common/conn_pool/conn_pool_base.cc

```diff
@@ -33,11 +33,44 @@ void ConnPoolImplBase::destructAllConnections() {
   dispatcher_.clearDeferredDeleteList();
 }
 
-void ConnPoolImplBase::tryCreateNewConnection() {
-  if (pending_requests_.size() <= connecting_request_capacity_) {
-    // There are already enough CONNECTING connections for the number
-    // of queued requests.
-    return;
+bool ConnPoolImplBase::shouldCreateNewConnection() const {
+  // The number of requests we want to be provisioned for is the number of
```

Comment:
nit: s/requests/streams here and below. Might be worth it to s/pending_requests/pending_streams, etc. as this is shared between TCP and HTTP (I think), but not a big deal and can be done later.